students/bplanica/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,210 @@
+            next(outfile) #skip header line
+            for row in outfile:
+                try:
+                    elements = row.split(",") #split by csv
+                    rental_id = elements[0].strip()
+                    user_id = elements[1].strip()
+                    product_id = elements[2].strip()
+                    rental_ip = {"_id":rental_id, "user_id":user_id, "product_id":product_id}
+                    rentals.insert_one(rental_ip)
+                    processed += 1
+                except (mongoerror.DuplicateKeyError, IndexError):
+                    logging.error("Import unsuccessful.")
+    except FileNotFoundError:
+        logging.error("File not found.")
+
+    after = record_count(rentals)
+    end = time.time()
+    logging.info("Rental: %s, %s, %s, %s", processed, before, after, end - start)
+    mod_queue.put((processed, before, after, end - start))
+
+
+def clear_data():
+    """drop all tables in the database"""
+    try:
+        DB.products.drop()
+        DB.customers.drop()
+        DB.rentals.drop()
+        logging.info("All tables have been dropped.")
+    except NameError:
+        logging.error("An Error has occurred; tables have not been dropped.")
+
+
+def show_available_products():
+    """Returns a Python dictionary of products listed as available with the following
+    fields: product_id, description, product_type, quantity_available."""
+    result = {}
+    for item in DB.products.find({"quantity_available":{"$gt":"0"}}):
+        logging.debug(item)
+        result[item['_id']] = {"description": item['description'],
+                               "product_type": item['product_type'],
+                               "quantity_available": item['quantity_available']}
+    return result
+
+
+def show_rentals(product_id):
+    """Returns a Python dictionary with the following user information from users that have
+    rented products matching product_id: user_id, name, address, phone_number, email."""
+    result = {}
+    for item in DB.rentals.find({"product_id":{"$eq":product_id}}):
+        user = item['user_id']
+        logging.debug(user)
+        for subitem in DB.customers.find({"_id":{"$eq":user}}):
+            logging.debug(subitem)
+            result[subitem['_id']] = {"name": subitem['name'], "address": subitem['address'],
+                                      "phone": subitem['phone'], "email": subitem['email']}
+    return result
+
+
+if __name__ == '__main__':
+    clear_data()
+    TOTAL_START = time.time()
+
+    QUEUE = Queue()
+    RESULTS = []
+    THREADS = []
+    THREADS.append(threading.Thread(target=import_product, args=("./", "products.csv", QUEUE)))
+    THREADS.append(threading.Thread(target=import_customer, args=("./", "customers.csv", QUEUE)))
+    THREADS.append(threading.Thread(target=import_rental, args=("./", "rentals.csv", QUEUE)))
+
+    for thread in THREADS:
+        thread.start()
+        RESULTS.append(QUEUE.get())
```

Comment:
Hi Bree, 

There's a problem with this approach: You're not waiting for the thread to finish and instead are assuming it has already finished by the time you call the _append_ operation. Instead, launch all your threads (the queue is thread-safe, so they'll be able to share it), then run another _for_ loop to _join_ each thread (this will ensure that all your threads have finished). Note that this means that when you run a _get_ operation on your queue, you will not to figure out which thread placed which data in the queue (you might add a string in the data each thread pushes or something else).

Let me know if this makes sense.

Regards,

Luis